File: js/create_new_users.js
var Promise = require('promise');
var twitter_helper = require ('./twitter_helper.js');
var _log = require('./log_helper.js')._log;
var child_process = require('child_process');
var util = require('util');
var settings = require('./settings.js');
var path = require("path");
var fs = require('fs');


var dm_id_file = path.join(settings.base_dir, '../data/dm_id.txt');
var dm_id_list = path.join(settings.base_dir, '../data/dm_id_list.txt');


var exec_cmd = function(cmd, success_msg) {
    return new Promise(function (fulfill, reject) {
        _log(util.format('++ cmd: %s', cmd));
        child_process.exec(cmd, function (error, stdout, stderr) {
            //_log(util.format('stdout: %s', stdout));
            if (error) {
                _log(util.format('XX: error: %j', error));
                _log(util.format('XX: stderr: %s', stderr));
                reject(error);
            }
            else if (error == null && success_msg != null) {
                _log(success_msg);
                fulfill(error);
            }
            else {
                fulfill(stdout);
            }
        });
    });
};

var create_user = function(username, password) {
    username = username.replace(/\W/g, '');
    password = password.replace(/\W/g, '');
    var create_user_script = path.join(settings.base_dir, 'bash/create_new_user.sh');
    var create_user_cmd = util.format('%s %s %s', create_user_script, username, password);
    return exec_cmd(create_user_cmd, util.format('++ @channel: successfully created user: %s', username));
};


var process_dm = function(dm) {
    console.log(util.format('screen_name: %s', dm['sender_screen_name']));
    console.log(util.format('message: %s', dm['text']));
    console.log(util.format('id: %s', dm['id']));
    // write id to a file
    var myRegexp = /(\S+)\s*\/\s*(\S+)/g;
    var match = myRegexp.exec(dm['text']);
    if (match) {
        var username = match[1];
        var password = match[2];
        console.log(util.format('++ create user: %s %s', username, password));
        var user_promise = create_user(username, password);
        // now tweet back at the user that their user was created
        user_promise.done(function(res) {
            twitter_helper.post_dm('OK', dm['sender_screen_name']);
            // get most recent tweet of correct format
            var tweets_promise = twitter_helper.get_timeline();
            tweets_promise.done(function(response) {
                var tweets = JSON.parse(response['body']);
                console.log(util.format('tweet: %j', Object.keys(tweets[0])));
                var host = '';
                var port = '';
                var index = 0;
                while (host == '') {
                    var tweet = tweets[index];
                    var myRegexp = /ssh \S+@(\S+) -p(\S+)/g;
                    var match = myRegexp.exec(tweet['text']);
                    if (match) {
                        host = match[1];
                        port = match[2];
                    }
                    index += 1;
                }
                twitter_helper.post_tweet(util.format('@%s ssh %s@%s -p%s', dm['sender_screen_name'], username, host, port));
            });
        }, function(err) {
            twitter_helper.post_dm('500: invalid username', dm['sender_screen_name']);
        });
        // ls /home
    }
    else {
        twitter_helper.post_dm('500: invalid username', dm['sender_screen_name']);
    }
};

var read_dm_id = function() {
    return new Promise(function (fulfill, reject) {
        if (fs.existsSync(dm_id_file)) {
            fs.readFile(dm_id_file, 'utf8', function (err, data) {
                if (err) {
                    reject(err);
                }
                fulfill(data);
            });
        }
        else {
            fulfill(0);
        }
    });
};


var find_already = function(dm_id) {
    var grep_cmd = util.format('grep -c %s %s', dm_id, dm_id_list);
    return new Promise(function (fulfill, reject) {
        child_process.exec(grep_cmd, function (error, stdout, stderr) {
            //_log(util.format('stdout: %s', stdout));
            if (error) {
                fulfill(0);
            }
            else {
                fulfill(stdout);
            }
        });
    });
};


var create_new_users = function() {
    _log('++ looking for new users to create');
    // read id from a file
    var last_dm_id_promise = read_dm_id();
    last_dm_id_promise.done(function(dm_id) {
        _log(util.format('last_dm_id: %s', dm_id));
        var dms_promise = twitter_helper.get_dms(dm_id);
        console.log('dms_promise: %j', dms_promise);
        dms_promise.done(function(response) {
            var dms = JSON.parse(response['body']);
            for (i = 0; i < dms.length; i++) {
                var dm = dms[i];
                if (i == 0) {
                   exec_cmd(util.format('echo "%s" > %s', dm['id'], dm_id_file));
                }
                var already_promise = find_already(dm['id']);
                already_promise.done(function(count) {
                    console.log(util.format('++ count: %s', count));
                    if (count == '0') {
                        exec_cmd(util.format('echo "%s" >> %s', dm['id'], dm_id_list));
                        _log(util.format('++ @channel reading dm with id: %s for user %s', dm['id'], dm['sender_screen_name']));
                        process_dm(dm);
                    }
                });
            }
        });
    });
};



if (require.main === module) {
    create_new_users();
}

File: oasis_python/create_new_users.py
import re
import os
import subprocess
import sys
import traceback

from pi_utilities.slack_helper import slack_notify_message
from pi_utilities.twitter_helper import TwitterHelper
from hello_settings import SECRETS_DICT, PROJECT_PATH

# initialize twitter
tw = TwitterHelper(access_token_key=SECRETS_DICT['TWITTER_ACCESS_TOKEN_KEY'],
                   access_token_secret=SECRETS_DICT['TWITTER_ACCESS_TOKEN_SECRET'],
                   consumer_key=SECRETS_DICT['TWITTER_CONSUMER_KEY'],
                   consumer_secret=SECRETS_DICT['TWITTER_CONSUMER_SECRET'])


# constants
DM_FILE_PATH = os.path.abspath(os.path.join(PROJECT_PATH, '../data/dm_id.txt'))
BOT_SCREEN_NAME = 'ss022001'


def get_last_dm_id():
    if not os.path.exists(DM_FILE_PATH):
        return 0
    with open(DM_FILE_PATH, 'r') as dm_file:
        dm_id = dm_file.read()
    return dm_id


def save_last_dm_id(dm_id):
    with open(DM_FILE_PATH, 'w') as dm_file:
        dm_file.write(str(dm_id))


def create_new_user(username, password):
    username = re.sub('[\W_]+', '', username)
    password = re.sub('[\W_]+', '', password)
    script_path = os.path.join(PROJECT_PATH, 'bash/create_new_user.sh')
    cmd = '{script} {username} {password}'.format(
        script=script_path,
        username=username,
        password=password
    )
    try:
        subprocess.check_output(cmd, shell=True, stderr=subprocess.STDOUT)
        return username
    except subprocess.CalledProcessError as e:
        sub_output = e.output
        raise Exception('XX: subprocess command failure: {}'.format(str(sub_output)))


def get_latest_host_and_port():
    tweets = tw.get_latest_tweets(screen_name=BOT_SCREEN_NAME, count=20)
    tweets.sort(key=lambda tweet: tweet['id'], reverse=True)
    for tweet in tweets:
        tweet_match = re.match('ssh \S+@(\S+) -p(\S+)', tweet['text'])
        if tweet_match:
            return {
                'host': tweet_match.group(1),
                'port': tweet_match.group(2)
            }


def process_dm(dm_text, dm_sender_screen_name, dm_id=''):
    dm_match = re.match('(\S+)\s*\/\s*(\S+)', dm_text)
    if not dm_match:
        raise Exception('dm message in invalid format')
    username = dm_match.group(1)
    password = dm_match.group(2)
    # check if there are any special characters in username or password
    if re.match('^[\w_]+$', username) is None:
        raise Exception('no special characters allowed in username')
    if re.match('^[\w_]+$', password) is None:
        raise Exception('no special characters allowed in password')
    # now create user
    create_new_user(username=username, password=password)
    # now dm user and tweet to them
    tw.send_dm(msg='OK', recipient=dm_sender_screen_name)
    # get the latest host and port
    latest_dict = get_latest_host_and_port()
    ssh_cmd = 'ssh {username}@{host} -p{port}'.format(
        username=username,
        host=latest_dict['host'],
        port=latest_dict['port']
    )
    tweet_text = '@{screen_name} {ssh_cmd}'.format(
        screen_name=dm_sender_screen_name,
        ssh_cmd=ssh_cmd
    )
    tw.post_tweet(tweet_text)
    tw.send_dm(msg=ssh_cmd, recipient=dm_sender_screen_name)
    slack_notify_message('++ @channel: created user {username} for screen name '
                         '@{screen_name} based on dm {dm_id}'.format(
        username=username,
        screen_name=dm_sender_screen_name,
        dm_id=dm_id
    ))


def create_new_users():
    slack_notify_message(message='++ looking for new users', channel_id='C19FPTANB')
    last_dm_id = get_last_dm_id()
    dms = tw.get_dms(since_id=last_dm_id)
    dms.sort(key=lambda dm: dm['id'])
    for dm in dms:
        save_last_dm_id(dm_id=dm['id'])
        try:
            process_dm(dm_text=dm['text'], dm_sender_screen_name=dm['sender_screen_name'], dm_id=dm['id'])
        except Exception as e:
            slack_notify_message('++ @channel: failed to process dm {dm_id} for user @{screen_name}: {error}'.format(
                dm_id=dm['id'],
                screen_name=dm['sender_screen_name'],
                error=e.message
            ))
            exc_type, exc_value, exc_traceback = sys.exc_info()
            formatted_lines = traceback.format_exc()
            slack_notify_message(formatted_lines)
            # send invalid username dm
            tw.send_dm(msg='500: {}'.format(e.message), recipient=dm['sender_screen_name'])


if __name__ == '__main__':
    create_new_users()
